lib/Runtime/PlatformAgnostic/IPlatformAgnosticResource.h

```diff
@@ -0,0 +1,63 @@
+//-------------------------------------------------------------------------------------------------------
+// Copyright (C) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE.txt file in the project root for full license information.
+//-------------------------------------------------------------------------------------------------------
+
+#ifndef RUNTIME_PLATFORM_AGNOSTIC_COMMON_IFINALIZABLE
+#define RUNTIME_PLATFORM_AGNOSTIC_COMMON_IFINALIZABLE
+
+namespace PlatformAgnostic
+{
+namespace Resource
+{
+    // Interface
+    class IPlatformAgnosticResource
+    {
+    public:
+        virtual void Cleanup() = 0;
+    };
+
+    template <typename T>
+    class AutoPtr
+    {
+    private:
+        const T *pointer;
+    public:
+        AutoPtr(const T *ptr) : pointer(ptr) {}
+        ~AutoPtr()
+        {
+            if (pointer != nullptr)
+            {
+                delete pointer;
+                pointer = nullptr;
+            }
+        }
+
+        void setPointer(const T *ptr)
+        {
+            pointer = ptr;
+        }
+    };
+
+    template <typename T>
+    class StringBufferAutoPtr
+    {
+    private:
+        const T *strBuf;
+    public:
+        StringBufferAutoPtr(const T *buffer) : strBuf(buffer) {}
+        ~StringBufferAutoPtr()
+        {
+            delete[] strBuf;
```

Comment:
I feel like these two classes should look exactly the same except `StringBufferAutoPtr` should be renamed to `AutoArrayPtr` and should use `delete[]` instead of `delete`

Reply:
Yeah, basically this is AutoArrayPtr in general (already does `delete[]` instead of `delete`). I could update this but I'm kind of meh on that. There's kind of a logical abstraction between arrays and strings in general parlance even though they are both pointers in practice in C++.